src/e2e/java/teammates/e2e/pageobjects/AppPage.java

```diff
@@ -0,0 +1,621 @@
+     * TODO: remove this method as it does not return necessary html anymore since frontend is generated by angular
+     */
+    public String getPageSource() {
+        return browser.driver.getPageSource();
+    }
+
+    public String getTitle() {
+        return browser.driver.getTitle();
+    }
+
+    public String getPageTitle() {
+        return browser.driver.findElement(By.tagName("h1")).getText();
+    }
+
+    public void click(By by) {
+        WebElement element = browser.driver.findElement(by);
+        click(element);
+    }
+
+    protected void click(WebElement element) {
+        executeScript("arguments[0].click();", element);
+    }
+
+    /**
+     * Simulates the clearing and sending of keys to an element.
+     *
+     * <p><b>Note:</b> This method is not the same as using {@link WebElement#clear} followed by {@link WebElement#sendKeys}.
+     * It avoids double firing of the {@code change} event which may occur when {@link WebElement#clear} is followed by
+     * {@link WebElement#sendKeys}.
+     *
+     * @see AppPage#clearWithoutEvents(WebElement)
+     */
+    private void clearAndSendKeys(WebElement element, CharSequence... keysToSend) {
+        Map<String, Object> result = clearWithoutEvents(element);
+        @SuppressWarnings("unchecked")
+        Map<String, String> errors = (Map<String, String>) result.get("errors");
+        if (errors != null) {
+            throw new InvalidElementStateException(errors.get("detail"));
+        }
+
+        element.sendKeys(keysToSend);
+    }
+
+    /**
+     * Clears any kind of editable element, but without firing the {@code change} event (unlike {@link WebElement#clear()}).
+     * Avoid using this method if {@link WebElement#clear()} meets the requirements as this method depends on implementation
+     * details.
+     */
+    private Map<String, Object> clearWithoutEvents(WebElement element) {
+        // This method is a close mirror of HtmlUnitWebElement#clear(), except that events are not handled. Note that
+        // HtmlUnitWebElement is mirrored as opposed to RemoteWebElement (which is used with actual browsers) for convenience
+        // and the implementation can differ.
+        checkNotNull(element);
+
+        // Adapted from ExpectedConditions#stalenessOf which forces a staleness check. This allows a meaningful
+        // StaleElementReferenceException to be thrown rather than just getting a boolean from ExpectedConditions.
+        element.isEnabled();
+
+        // Fail safe in case the implementation of staleness checks is changed
+        if (isExpectedCondition(ExpectedConditions.stalenessOf(element))) {
+            throw new AssertionError(
+                    "Element is stale but should have been caught earlier by element.isEnabled().");
+        }
+
+        @SuppressWarnings("unchecked")
+        Map<String, Object> result = (Map<String, Object>) executeScript(
+                "const element = arguments[0];"
```

Comment:
For long (> 3 lines) scripts like this (not just here), let's extract them out. You can check here for example:

https://github.com/TEAMMATES/teammates/blob/f966efc68b19a39b58dccc5d48dccea397996b76/src/e2e/resources/scripts/waitForPageLoad.js#L1-L26
https://github.com/TEAMMATES/teammates/blob/f966efc68b19a39b58dccc5d48dccea397996b76/src/e2e/java/teammates/e2e/pageobjects/Browser.java#L24-L33
https://github.com/TEAMMATES/teammates/blob/f966efc68b19a39b58dccc5d48dccea397996b76/src/e2e/java/teammates/e2e/pageobjects/Browser.java#L80-L85